### PR TITLE
Add numexpr

### PR DESCRIPTION
**Pull request by ocefpaf**
@msarahan I am trying to resurrect #547. Let me know if this is OK.

My goal is to get a `pytables` build but we can only do `numexpr`+ `mkl` from `defaults` for now b/c we don't have `numexpr` on conda-forge.

### CODE REVIEW

**Comment by jakirkham** [comment]
> we can only do `mkl` for now b/c we don't have `numexpr` on conda-forge

So that is an interesting point worthy of discussion. How do you plan to support `mkl`? We don't have the headers.

**Comment by ocefpaf** [comment]
> How do you plan to support `mkl`?

You misread me above. 

> My goal is to get a `pytables` build

There is no need for the headers as everything I need for `pytables` is already on `defaults`.

File: appveyor.yml [hunk 1]
```diff
@@ -88,6 +88,7 @@ install:
     - cmd: conda install --yes --quiet obvious-ci conda-build-all
     - cmd: conda install --yes --quiet conda-forge-build-setup
     - cmd: run_conda_forge_build_setup
+    - cmd: conda install --yes --quiet conda-build=1.21.5
```
Comment (ocefpaf) [inline appveyor.yml]: We need to remove this before merging the PR. Just wanted to make sure this works.

PS: This is not needed in the feedstock.

**Comment by jakirkham** [comment]
> > How do you plan to support `mkl`?

> You misread me above. 

Ok, so do you mean you are going to try to support OpenBLAS then?

File: recipes/numexpr/meta.yaml [hunk 2]
```diff
@@ -0,0 +1,49 @@
+{% set version="2.6.1" %}
```
Comment (jakirkham) [inline recipes/numexpr/meta.yaml]: `{% set variant = "openblas" %}`

File: recipes/numexpr/meta.yaml [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+source:
+    fn: numexpr-{{ version}}.tar.gz
+    url: https://github.com/pydata/numexpr/archive/v{{ version }}.tar.gz
+    sha256: e92c83d066fa8da63864d69b5f218287cc31437ae844db77390f2183123aab22
+
+build:
+    number: 0
+    script: python setup.py install --single-version-externally-managed --record=record.txt
+    features:  # [not win]
+        - blas_openblas  # [not win ]
```
Comment (jakirkham) [inline recipes/numexpr/meta.yaml]: Please change to...

`- blas_{{ variant }}`

File: recipes/numexpr/meta.yaml [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+    name: numexpr
+    version: {{ version }}
+
+source:
+    fn: numexpr-{{ version}}.tar.gz
+    url: https://github.com/pydata/numexpr/archive/v{{ version }}.tar.gz
+    sha256: e92c83d066fa8da63864d69b5f218287cc31437ae844db77390f2183123aab22
+
+build:
+    number: 0
```
Comment (jakirkham) [inline recipes/numexpr/meta.yaml]: Please set to `200`.

File: recipes/numexpr/meta.yaml [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+    script: python setup.py install --single-version-externally-managed --record=record.txt
+    features:  # [not win]
+        - blas_openblas  # [not win ]
+
+requirements:
+    build:
+        - python
+        - setuptools
+        - numpy  x.x
```
Comment (jakirkham) [inline recipes/numexpr/meta.yaml]: Please add the following.

```
    - blas 1.1 {{ variant }}
    - openblas 0.2.18*
```

File: recipes/numexpr/meta.yaml [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+    run:
+        - python
+        - numpy  x.x
```
Comment (jakirkham) [inline recipes/numexpr/meta.yaml]: Please add the following here too.

```
    - blas 1.1 {{ variant }}
    - openblas 0.2.18*
```

**Comment by jakirkham** [comment]
Add some changes based on how we handle BLAS at this point. The details can be found this [hackpad]( https://conda-forge.hackpad.com/BLAS-Numpy-Friends-86De62hoNdT ).

**Comment by jakirkham** [comment]
I'm not really sure what we can do about Windows here as we still don't have a BLAS to use there.

**Comment by ocefpaf** [comment]
@msarahan the Windows issue is gone! Thanks!!

> Ok, so do you mean you are going to try to support OpenBLAS then?

No. `numexpr` can use `mlk`, but I don't see anything in the docs about `openblas`. (And I would not use a single threaded openblas anyways, if you read `numexpr` docs you will see that it is only more efficient with the multi-threading offered in `mlk`.)

@jakirkham can you explain why you need all those changes regarding `openblas`? I believe you are missing the point here. The only reason `conda-forge` needs `numexpr` is to avoid getting `mkl` installed (`mkl` comes with `numexpr` and `numpy` from `defaults`) at `pytables` compilation. Due to channel preference we get `numexpr` with `mkl`, but `numpy` without and that causes `pytables` build to fail due to missing symbols. See https://circleci.com/gh/ocefpaf/staged-recipes/190

**Comment by jakirkham** [comment]
@msarahan, pointed out in the [original PR]( https://github.com/conda-forge/staged-recipes/pull/547#issue-153111989 ) this needs a BLAS (hence the comments above). However, as I'm looking at `numexpr`'s source code, I'm inclined to agree with you that it doesn't seem to. Maybe @msarahan could explain that comment in more detail.

**Comment by msarahan** [comment]
This explains it pretty well: https://github.com/pydata/numexpr/wiki/Numexpr-Users-Guide#enabling-intels-vml-support

Not necessary, especially not for a first pass.

**Comment by jakirkham** [comment]
Ok, so it is not that it requires some BLAS. It is just that it optionally uses Intel's VML if we build with.

Then I agree @ocefpaf, we do not need OpenBLAS support for this whatsoever. I withdraw my comments related to such changes.

File: recipes/numexpr/meta.yaml (hunk 4 above)
Comment (jakirkham) [inline recipes/numexpr/meta.yaml]: Retracted.

File: recipes/numexpr/meta.yaml (hunk 2 above)
Comment (jakirkham) [inline recipes/numexpr/meta.yaml]: Retracted.

File: recipes/numexpr/meta.yaml [hunk 7]
```diff
@@ -0,0 +1,47 @@
+{% set version="2.6.1" %}
+
+package:
+    name: numexpr
+    version: {{ version }}
+
+source:
+    fn: numexpr-{{ version}}.tar.gz
```
Comment (jakirkham) [inline recipes/numexpr/meta.yaml]: nit: `{{ version}}` -> `{{ version }}`

File: recipes/numexpr/meta.yaml [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+    url: https://github.com/pydata/numexpr/archive/v{{ version }}.tar.gz
+    sha256: e92c83d066fa8da63864d69b5f218287cc31437ae844db77390f2183123aab22
+
+build:
+    number: 0
+    script: python setup.py install --single-version-externally-managed --record=record.txt
+
+requirements:
+    build:
```
Comment (jakirkham) [inline recipes/numexpr/meta.yaml]: Please add `toolchain`.

Comment (ocefpaf) [inline recipes/numexpr/meta.yaml]: Can you specify what flags are relevant for this case?

I am not adding the `toolchain` in any of the feedstocks I maintain for the following reasons:

- those flags are not needed all the time\*
- the extra flags may cause unexpected issues like the one we faced with `gdal` (ping @kmuehlbauer) making debuging hard due to the unknonwn changes
- the `toolchain` is not versioned in any `feedstock`, but it is constantly updated breaking build reproducibility
- even if we start versioning them we are adding an extra layer that makes it difficult to inspect and third party using our recipes are hurt b/c of that

\* The only cases I see the where some flags are needed is to use `clang` instead of `gcc`. If we have a `gcc` recipe in the future and adopt it everywhere the `toolchain` won't be needed at all. But that is another topic for the future.

**Comment by ocefpaf** [comment]
@msarahan any comments? Does this looks good to you?

**Comment by msarahan** [comment]
Looks fine to me, but I would like to understand whether we need flags to use clang.  I think most of the rest of our system uses it on mac, right?

I agree with all of @ocefpaf's concern over toolchain, but also see toolchain as the only scalable way to handle this kind of thing right now.  I think that the build customization stuff I'm working on for conda-build might help this situation, because it will at least "bake in" the customization stuff into the final recipe, which would then be available for inspection.  Is there a way for the toolchain package to do something in meantime to provide similar "baked-in" readily accessible information about any settings it made?  Are there overrides for toolchain, such that each of its settings can be overridden as necessary?

**Comment by ocefpaf** [comment]
> but also see toolchain as the only scalable way to handle this kind of thing right now.

I know that @pelson has a few ideas.

I agree that copying and pasting the flags is not "very scalable," but I wonder how many places those flags are really needed. (Specially if we move to `gcc` everywhere once we had it packaged.)

> Looks fine to me, but I would like to understand whether we need flags to use clang.

That I can answer. See https://travis-ci.org/conda-forge/staged-recipes/jobs/146985729#L1464

There is no `gcc` b/c we remove the `homebrew` version in the `OS X` image. This **is** using `clang` as everything else (unless we add `gcc` from `defaults` manually). So the flags are not needed. So far I only see the need fir the flags when there is `c++` code that needs the `libc++` vs `libstdc++` situation sorted out. I cannot see any `c++` in `numexpr`, so I did not add those.

If there is `c++` in `numexpr` I will gladly add them.

**Comment by msarahan** [comment]
Well, if the output here has no libc++/libstdc++, then I don't see an urgent need to add toolchain.  I welcome comment from @pelson on any other ideas he has.  I think the final nail in the coffin for me would be output from ```conda inspect linkages``` on the mac package.

I see some wisdom in adding toolchain to any compiled library, for the sake of some "future-proofing."  However, given the troubles that some have encountered because of this approach, I am also OK with not including it here, where we have good reason for saying that it is not currently useful.  It just means that we may need to revisit this if toolchain ever does include something that changes this.

**Comment by ocefpaf** [comment]
> conda inspect linkages on the mac package.

@msarahan I would love to have `conda inspect linkages` as a default test in all feedstocks, but it seems that `conda inspect linkages` is still broken on `OS X`.

**Comment by msarahan** [comment]
Ugh.  Well, the good news is that I'm making sure it works as part of the PR creating the Python API, because I didn't want that to break anything...

**Comment by ocefpaf** [comment]
> Ugh.  Well, the good news is that I'm making sure it works as part of the PR creating the Python API, because I didn't want that to break anything...

Thanks! That is probably one of the most useful feature for building conda packages.

**Comment by jakirkham** [comment]
> I cannot see any `c++` in `numexpr`, so I did not add those.

Here is some C++ [code]( https://github.com/pydata/numexpr/blob/v2.6.1/numexpr/interpreter.cpp#L15 ). There is [more]( https://github.com/pydata/numexpr/search?l=cpp ).

> Well, if the output here has no libc++/libstdc++, then I don't see an urgent need to add toolchain.

It does. See this [line]( https://travis-ci.org/conda-forge/staged-recipes/jobs/147037230#L2317 ).

File: recipes/numexpr/meta.yaml (hunk 8 above)
Comment (jakirkham) [inline recipes/numexpr/meta.yaml]: I've tried to respond to your points below. Though I'm not going to have time to debate this at present. If this is so controversial, maybe we will have to save it for the next meeting.

> those flags are not needed all the time

Unfortunately, they are needed fairly often when it comes to compiling. Particularly on OS X. As stated before, I'm not interested in setting up custom flags for everything. I tried it for awhile and it was exhausting. Also it was impossible to keep up with the rate things were being added. Hence why some things ended up using `stdlibc++` thus resulting in breakage.

> the extra flags may cause unexpected issues like the one we faced with `gdal`...making debuging hard due to the unknonwn changes

While I understand there are some special like GDAL where this has had issues (though I still need to catch up on exactly what the issues were in that case), they are pretty rare and they also tended to be cases that we already had substantial issues getting built in the first place (again GDAL). I can guarantee that `numexpr` is not one of these. I built it no problem with the `toolchain` locally. In any event, the suggestion that we reproduce and maintain these flag everywhere because there are a few cases needing special attention sounds inappropriate for the problem, unstainable in the long run, and even more problematic when mixing things as we have no real guarantees.

> the `toolchain` is not versioned in any `feedstock`, but it is constantly updated breaking build reproducibility

If it is productive for you to pin it to a version, go for it. However, the changes made of late have been relatively subtle.

> even if we start versioning them we are adding an extra layer that makes it difficult to inspect and third party using our recipes are hurt b/c of that

Not sure I follow. The few lines of code involved is open, simple, and easily gotten from `conda-forge`.

> The only cases I see the where some flags are needed is to use `clang` instead of `gcc`. If we have a `gcc` recipe in the future and adopt it everywhere the `toolchain` won't be needed at all.

Not so sure about that. See this [comment]( https://github.com/conda-forge/docker-images/pull/22#issuecomment-233403402 ). It looks like we will be playing the same game with ABI compatibility. Just we will have to do it on Linux too.

Not to mention, that you are then volunteering to update these flags in 100s of places when we switch. I certainly would not want to do that.

> But that is another topic for the future.

Agreed.

**Comment by ocefpaf** [comment]
> It does. See this line.

That is why I added the test. It would be nice if someone that owns a Mac could step up and help with the tests here.

I am curious to the extend of the issue here since we have many extensions in a similar situation that are not using those flags nor the `toolchain` and they are all fine. Maybe that the scope of the problem is limited and none of these are affected. Sadly it is hard to find proper information.

I am tempted to merge this as is and ask for people to test it. If it fails I already have the proper fix in line.

File: recipes/numexpr/meta.yaml (hunk 8 above)
Comment (ocefpaf) [inline recipes/numexpr/meta.yaml]: >
> In recipes/numexpr/meta.yaml
> <https://github.com/conda-forge/staged-recipes/pull/1133#discussion_r71996736>
> :
>
> > +
> > +package:
> > +    name: numexpr
> > +    version: {{ version }}
> > +
> > +source:
> > +    fn: numexpr-{{ version}}.tar.gz
> > +    url: https://github.com/pydata/numexpr/archive/v{{ version }}.tar.gz
> > +    sha256: e92c83d066fa8da63864d69b5f218287cc31437ae844db77390f2183123aab22
> > +
> > +build:
> > +    number: 0
> > +    script: python setup.py install --single-version-externally-managed --record=record.txt
> > +
> > +requirements:
> > +    build:
>
> I've tried to respond to your points below. Though I'm not going to have
> time to debate this at present. If this is so controversial, maybe we will
> have to save it for the next meeting.
>

It is already in the meeting agenda.

**Comment by jakirkham** [comment]
> I am tempted to merge this as is and ask for people to test it. If it fails I already have the proper fix in line.

I'm not sure why a self-merge is being consider appropriate when something is still needing to be addressed. Especially when it is still something under discussion.

cc @pelson

**Comment by ocefpaf** [comment]
> I'm not sure why a self-merge

Why do you think I would do that? When I said I am tempted, but I always respected time and discussion here. In fact I am constantly attacked for asking other for time and patience.

Please reduce the noise here otherwise we won't get important feedback from others.

**Comment by dharhas** [comment]
> That is why I added the test. It would be nice if someone that owns a Mac could step up and help with the tests here.

@ocefpaf do you still need tests run on OS X? I can help if needed, I will need some guidance on how to build and test conda-forge recipes locally.

What else needs to be done? This pr is holding up #1134 which is holding up https://github.com/conda-forge/rasterio-feedstock/pull/13 which we cannot install gdal on any of our systems right now and I'm motivated to get this sorted out...

**Comment by ocefpaf** [comment]
> @ocefpaf do you still need tests run on OS X? I can help if needed, I will need some guidance on how to build and test conda-forge recipes locally.

Thanks for offering help @dharhas, but there is nothing to do. All the original unit-test passes on both `Linux` and `OS X` are passing here. Many other packages have the same linking "issue" above as I listed in https://github.com/conda-forge/staged-recipes/pull/1133#issuecomment-234806680 and none of those use nor need the `toolchain` nor presented the "expected" failures.

Adding the `toolchain` hinders some important build flags from the end users and many at IOOS are against that since that prevents them from easily debugging problems.

Not to mention that the `toolchain` is not versioned in any `feedstock` breaking build reproducibility as we don't known what flags from what version were used for the builds. It is also important to not that the `toolchain` is only needed in a few cases and adding it everywhere is not needed at all. And last, but not least, we should be focusing our efforts on a `gcc` package to standardize our builds from the ground up and no on adding workarounds.

I am sorry. I know this is a drag. I already added a topic for the meeting and I hope common sense will prevail.

Like I said in https://github.com/conda-forge/staged-recipes/pull/1133#issuecomment-234806431 we should merge this and move forward. If there is a real problem we can easily fix it by adding the necessary build flags in the `feedstock` itself and securing the `feedstock` history and reproducibility

**Comment by dharhas** [comment]
@ocefpaf @jakirkham @pelson @msarahan 

Sorry to be a bother, but it looks like this conversation has stalled. I really appreciate everything y'all are doing with conda-forge and the thoughtfulness being put into making it robust and viable, but this is causing severe issues in our group. We are getting ready for internal demos and a product release and need to be able to install reliably cross platform. This is holding up multiple downstream packages and unless this gets resolved and merged we cannot move forward.

Can we follow @ocefpaf suggestion and merge this and deal with the 'toolchain/no toolchain' question separately and more holistically at a future point?

**Comment by patricksnape** [comment]
@jakirkham @ocefpaf Just to try and arbitrate here - given @ocefpaf is correct in pointing out that many recipes that perhaps *should* have `toolchain` currently do not - **perhaps we should just merge this and hash out this issue during the meeting?** 

From my end there are two major issues that I think are important (both for and against):
  - `toolchain` is not versioned, indicated in the build string, or easily overridable
  - `toolchain` is a necessary evil due to things like ABI compatibility - see [this PR](https://github.com/conda-forge/staged-recipes/pull/1094) where setting `MACOSX_DEPLOYMENT_TARGET` may be necessary for **all** OSX builds on conda-forge.

**Comment by msarahan** [comment]
I agree with @patricksnape, but think that this deserves a special meeting of its own.

**Comment by ocefpaf** [comment]
>  - `toolchain` is not versioned, indicated in the build string, or easily overridable

Yep. That is my number one concern. (Although the interference effect that `toolchain` has over some of the builds, like gdal and others, it not desirable defeating te purpose of all controling these flags in one place and making debugging extremely hard.)

  - `toolchain` is a necessary evil due to things like ABI compatibility - see [this PR](https://github.com/conda-forge/staged-recipes/pull/1094) where setting `MACOSX_DEPLOYMENT_TARGET` may be necessary for **all** OSX builds on conda-forge.

Once we settle on what flags are need then we should actually think about were they should live. Unfortunately they were implemented in a "works for me" kind of way at te cost of the stability of the GIS stack over the course of 1.5 months.

I am still against a `toochain` as we are preventing some users to compile software breaking, consciously of course, those flags. I can see may places where I'd prefer to use `gcc` and have `libgomp` available. Note that if I do that on an "end" package (something that won't be linking to anything else) then the issues are not available and nothing gets broken. Asking for the use of the `toolchain` will act like a virus that won't allows us to do that.

TL;DR I prefer to have some freedom and to set the flags explicitly, when they are really needed.

(I actually prefer to work on a `gcc` package and ditch `clang` and the `toolchain` altogether, but that is another problem for another day.)

**Comment by patricksnape** [comment]
But the `gcc` package doesn't fix the `MACOSX_DEPLOYMENT_TARGET` issue (at least as far as I understand). I'm not necessarily fighting for `toolchain`, but I understand it's existence and I definitely think that being pragmatic and setting this flags *somewhere* is a *good* idea. 

Still, this issue is quickly spiralling out of control - I think we should schedule a meeting for this and move forward with the availability of this package (by merging).

**Comment by ocefpaf** [comment]
> Still, this issue is quickly spiralling out of control - I think we should schedule a meeting for this and move forward with the availability of this package (by merging).

Agreed.

> But the `gcc` package doesn't fix the `MACOSX_DEPLOYMENT_TARGET` issue (at least as far as I understand). I'm not necessarily fighting for `toolchain`, but I understand it's existence and I definitely think that being pragmatic and setting these flags *somewhere* is a *good* idea. 

Indeed that is another discussion.

**Comment by dharhas** [comment]
@msarahan, @patricksnape @ocefpaf Looks like the consensus is the merge and move on. Can somebody who has merge rights please do the honors or at least say that a self merge by @ocefpaf is okay. I don't want to sound like I'm whining (unless it will help, and in that case I'm definitely whining :) ) but the entire geospatial stack is currently broken and myself and a lot of others need it to not be broken.

**Comment by msarahan** [comment]
I will merge once we have a meeting scheduled. Can someone post a doodle to the gitter channel? I'm traveling and don't want to fight my phone to do so.

**Comment by dharhas** [comment]
It has been almost a week. Any progress on merge plans for this?

**Comment by dharhas** [comment]
@msarahan, @patricksnape @ocefpaf @jakirkham 

Sorry to beat a dead horse but have y'all met, is this going to be merged? If not I need to start looking for workarounds for the GIS stack.

**Comment by msarahan** [comment]
Sorry.  I'm going to merge this as-is.  If there are *any* problems, we can try the addition of toolset.  I'm really sorry that this has gotten so hung up.  I hope I'll have time to work on our gcc-based compiler stack soon, so that this will no longer be an issue.

**Comment by dharhas** [comment]
Thanks!

**Comment by ocefpaf** [comment]
>  I hope I'll have time to work on our gcc-based compiler stack soon

:heavy_plus_sign: :100:
